src/health.rs

```diff
@@ -0,0 +1,307 @@
+pub struct HealthCheckResult {
+    /// Overall health status of client service.
+    /// `READY`, `NOT_READY`, or `UNKNOWN`.
+    /// TODO: We potentially want to expand this to distinguish `Live` and `Ready to serve` statuses.
+    pub health_status: HealthStatus,
+    /// Response code of the latest health check request.
+    /// This should be omitted on serialization if the health check was successful (when the response is `HTTP 200 OK` or `gRPC 0 OK`).
+    pub response_code: ClientCode,
+    /// Optional reason for the health check result status being `Not ready to serve` or `Unknown`.
+    /// May be omitted overall if the health check was successful.
+    pub reason: Option<String>,
+}
+
+/// A cache to hold the latest health check results for each client service.
+/// Orchestrator has a reference-counted mutex-protected instance of this cache.
+#[derive(Debug, Clone, Default, Serialize)]
+pub struct HealthCheckCache {
+    pub detectors: HashMap<String, HealthCheckResult>,
+    pub chunkers: HashMap<String, HealthCheckResult>,
+    pub generation: HashMap<String, HealthCheckResult>,
+}
+
+/// Response for the readiness probe endpoint that holds a serialized cache of health check results for each client service.
+#[derive(Debug, Clone, Serialize)]
+pub struct HealthProbeResponse {
+    pub services: HealthCheckCache,
+}
+
+/// Probe query param for the readiness probe endpoint.
+#[derive(Debug, Clone, Deserialize)]
+pub struct ReadyCheckParams {
+    /// Whether to probe the service for readiness or just return the cached health status.
+    #[serde(default)]
+    pub probe: bool,
+}
+
+impl HealthCheckResult {
+    pub fn is_ready(&self) -> bool {
+        matches!(self.health_status, HealthStatus::Ready)
+    }
+
+    pub fn is_not_ready(&self) -> bool {
+        matches!(self.health_status, HealthStatus::NotReady)
+    }
+
+    pub fn is_unknown(&self) -> bool {
+        matches!(self.health_status, HealthStatus::Unknown)
+    }
+
+    pub fn reason_from_health_check_response(response: &HealthCheckResponse) -> Option<String> {
+        match response.status {
+            0 => Some(
+                "gRPC serving status UNKNOWN: Service's health is unexpectedly unknown".to_string(),
+            ),
+            1 => None,
+            2 => Some(
+                "gRPC serving status NOT_SERVING: Service is not ready to serve requests"
+                    .to_string(),
+            ),
+            3 => Some(
+                "gRPC serving status SERVICE_UNKNOWN: Service's heath is currently unknown"
+                    .to_string(),
+            ),
+            _ => {
+                error!(
+                    "Unexpected gRPC health check response status: {}",
+                    response.status
+                );
+                Some(format!(
+                    "Unexpected gRPC health check response status: {}",
+                    response.status
+                ))
+            }
+        }
+    }
+}
```

Comment:
These descriptions feel redundant, aren't the status strings such as UNKNOWN, NOT_SERVING, etc.  self-explanatory?

Reply:
Regardless of whether we use `HEALTHY | UNHEALTHY | UNKNOWN` or just `HEALTHY | UNHEALTHY`, we can preserve the original gRPC status in non-healthy cases as part of the reason message for debugging, where gRPC health status is of `READY | NOT_READY | SERVICE_UNKNOWN | UNKNOWN`. I think this is something nice to preserve, although if these are too wordy maybe simpler messages like just `from gRPC serving_status: SERVICE_UNKNOWN`, omitting the redundant `Service's health is currently unknown`?